src/odbc_connection.cpp

```diff
@@ -2091,6 +2093,427 @@ Napi::Value ODBCConnection::GetInfo(const Napi::Env env, const SQLUSMALLINT opti
+        return;
+      }
+
+      return_code = prepare_for_fetch(data);
+      bool alloc_error = false;
+      return_code =
+      fetch_all_and_store
+      (
+        data,
+        false,
+        &alloc_error
+      );
+      if (alloc_error)
+      {
+        SetError("[odbc] Error allocating or reallocating memory when fetching data. No ODBC error information available.\0");
+        return;
+      }
+      if (!SQL_SUCCEEDED(return_code)) {
+        this->errors = GetODBCErrors(SQL_HANDLE_STMT, data->hstmt);
+        SetError("[odbc] Error retrieving table information results set\0");
+        return;
+      }
+    }
+
+    void OnOK() {
+
+      Napi::Env env = Env();
+      Napi::HandleScope scope(env);
+
+      std::vector<napi_value> callbackArguments;
+
+      callbackArguments.push_back(env.Null());
+
+      Napi::Array empty = Napi::Array::New(env);
+      Napi::Array rows = process_data_for_napi(env, data, empty);
+      callbackArguments.push_back(rows);
+
+      Callback().Call(callbackArguments);
+    }
+
+  public:
+
+    PrimaryKeysAsyncWorker(ODBCConnection *odbcConnectionObject, StatementData *data, Napi::Function& callback) : ODBCAsyncWorker(callback),
+      odbcConnectionObject(odbcConnectionObject),
+      data(data) {}
+
+    ~PrimaryKeysAsyncWorker() {
+      delete data;
+      data = NULL;
+    }
+};
+
+//
+//  ODBCConnection::PrimaryKeys
+//
+//    Description: Returns the columns names that make up the primary key for a
+//                 table.
+//
+//    Parameters:
+//      const Napi::CallbackInfo& info:
+//        The information passed from the JavaSript environment, including the
+//        function arguments for 'tables'.
+//
+//        info[0]: String: catalog
+//        info[1]: String: schema
+//        info[2]: String: table
+//        info[3]: Function: callback function:
+//            function(error, result)
+//              error: An error object if there was a database issue
+//              result: The ODBCResult
+//
+//    Return:
+//      Napi::Value:
+//        Undefined (results returned in callback)
+//
+Napi::Value ODBCConnection::PrimaryKeys(const Napi::CallbackInfo& info) {
+
+  Napi::Env env = info.Env();
+  Napi::HandleScope scope(env);
+
+  if (info.Length() != 4) {
+    Napi::TypeError::New(env, "primaryKeys() function takes 4 arguments.").ThrowAsJavaScriptException();
+  }
+
+  Napi::Function callback;
+  StatementData* data                      = new StatementData();
+                 data->henv                = this->hENV;
+                 data->hdbc                = this->hDBC;
+                 data->fetch_array         = this->connectionOptions.fetchArray;
+                 data->maxColumnNameLength = this->getInfoResults.max_column_name_length;
+                 data->get_data_supports   = this->getInfoResults.sql_get_data_supports;
+  // Napi doesn't have LowMemoryNotification like NAN did. Throw standard error.
+  if (!data) {
```

Comment:
This needs to be _before_ you try assigning `data->`. Also, unless you use [new (std::nothrow)](https://cplusplus.com/reference/new/nothrow/) the `new StatementData()` will _never_ return NULL, but instead will throw an exception and this null check will never work right.

Reply:
This is an issue not just in these new functions, but in all the "query" functions in this class. I won't fix them all here as to not pollute the commit, but will fix them up.